blockchain/sync_manager/src/sync.rs

```diff
@@ -122,6 +146,141 @@ impl<'a> Syncer<'a> {
+                    if *sequence != msg.sequence() {
+                        return Err(Error::Validation("Sequences are not equal".to_string()));
+                    }
+
+                    // sufficient funds check
+                    if balance < &msg.required_funds() {
+                        return Err(Error::Validation(
+                            "Insufficient funds for message execution".to_string(),
+                        ));
+                    }
+                    // update balance and increment sequence by 1
+                    MsgMetaData {
+                        balance: balance - msg.required_funds(),
+                        sequence: sequence + 1,
+                    }
+                }
+                // sequence is not found, insert sequence and balance with address as key
+                _ => {
+                    let actor = tree.get_actor(msg.from());
+                    if let Some(act) = actor {
+                        MsgMetaData {
+                            sequence: *act.sequence(),
+                            balance: act.balance().clone(),
+                        }
+                    } else {
+                        return Err(Error::State(
+                            "Could not retrieve actor from state tree".to_string(),
+                        ));
+                    }
+                }
+            };
+            msg_meta_data
+                .metadata
+                .insert(msg.from().clone(), updated_state);
+            Ok(())
+        }
+
+        let mut msg_meta_data = MsgCheck::new();
+        // TODO retrieve tipset state and load state tree
+        // temporary
+        let tree = HamtStateTree::default();
+        // loop through bls messages and check msg validity
+        for m in block.bls_msgs() {
+            check_msg(m.clone(), &mut msg_meta_data, &tree)?;
+        }
+        // loop through secp messages and check msg validity and signature
+        for m in block.secp_msgs() {
+            check_msg(m.clone(), &mut msg_meta_data, &tree)?;
+            // signature validation
+            if !is_valid_signature(&m.cid()?.to_bytes(), m.from(), m.signature()) {
+                return Err(Error::Validation(
+                    "Message signature is not valid".to_string(),
+                ));
+            }
+        }
+        // validate message root from header matches message root
+        let sm_root = self.compute_msg_data(&block)?;
+        if block.to_header().messages() != &sm_root {
+            return Err(Error::InvalidRoots);
+        }
+
+        Ok(())
+    }
+
+    /// Validates block semantically according to https://github.com/filecoin-project/specs/blob/6ab401c0b92efb6420c6e198ec387cf56dc86057/validation.md
+    pub fn validate(&self, block: Block) -> Result<(), Error> {
+        // get header from full block
+        let header = block.to_header();
+        let base_tipset = self.load_fts(TipSetKeys::new((*header.parents().cids).to_vec()))?;
```

Comment:
```suggestion
        let base_tipset = self.load_fts(TipSetKeys::new(header.parents().cids.clone()))?;
```
Just slightly confusing